File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/check_overlap.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval() : start(-1), end(-1) {}

  Interval(int start, int end) : start(start), end(end) {}
};

// O(nlogn) / O(1)
bool has_overlap(vector<Interval> &a) {
  sort(a.begin(), a.end(),
       [](auto &e1, auto &e2) { return e1.start <= e2.start; });

  int start = a[0].start;
  int end = a[0].end;

  for (int i = 1; i < a.size(); i++) {
    if (a[i].start <= end) {
      return true;
    } else {
      start = a[i].start;
      end = a[i].end;
    }
  }

  return false;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, start, end;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a(n);
    for (int i = 0; i < n; i++) {
      cin >> start >> end;
      a[i] = Interval(start, end);
    }

    cout << (has_overlap(a) ? "true" : "false") << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/CheckOverlap.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class CheckOverlap {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        System.out.println(checkOverlap(a));
      }
    }
  }

  // O(nlogn) / O(1)
  public static boolean checkOverlap(List<Interval> a) {
    a.sort((x, y) -> Integer.compare(x.start, y.start));

    BiFunction<Interval, Interval, Boolean> hasOverlap = (x, y) -> {
      return (x.start >= y.start && x.start <= y.end) ||
          (y.start >= x.start && y.start <= x.end);
    };

    for (int i = 0; i < a.size() - 1; i++) {
      if (hasOverlap.apply(a.get(i), a.get(i + 1))) {
        return true;
      }
    }

    return false;
  }
}